syntax = "proto3";

service AccountService {
  rpc GetAccountByName(GetAccountRequest) returns (GetAccountResponse);
  rpc SaveCurrentAccount(SaveAccountRequest) returns (SuccessMessage);
  rpc CreateNewAccount(CreateAccountRequest) returns (GetAccountResponse);
}
service UserService {
  rpc AddUser(UserRequest) returns (UserResponse);
  rpc GetUsers(Empty) returns (UserListResponse);
  rpc UpdateUser(UserRequest) returns (UpdateMessage);
}

service StatisticsService {
  rpc GetCurrentAccountStatistics (AccountRequest) returns (AccountStatisticsResponse);
  rpc UpdateAccountStatistics (UpdateAccountRequest) returns (UpdateAccountResponse);
}

message Account {
  string name = 1;
  repeated Item incomes = 3;
  repeated Item expenses = 4;
  Saving saving = 5;
  string note =6;
}

message Item {
  string title = 1;
  string amount = 2; // this is the type of BigDecimal
  Currency currency = 3;
  TimePeriod period = 4;
}

enum Currency {
  USD = 0;
  EUR = 1;
  RUB = 2;
}

enum TimePeriod {
  YEAR = 0;
  QUARTER = 1;
  MONTH = 2;
  DAY = 3;
  HOUR = 4;
}

message Saving {
  string amount = 1; // this is the type of BigDecimal
  Currency currency = 2;
  string interest = 3; // this is the type of BigDecimal
  bool deposit = 4;
  bool capitalization = 5;
}

message GetAccountRequest {
  string name = 1;
}

message GetAccountResponse {
  Account account = 1;
}

message SaveAccountRequest {
  string accountName = 1;
  repeated Item incomes = 2;
  repeated Item expenses = 3;
  Saving saving = 4;
}

message CreateAccountRequest {
  string username = 1;
  string password = 2;
}

message SuccessMessage{
  string successMessage = 1;
}

message User {
  string username = 1;
  string password = 2;
}

message UserRequest {
  User user = 1;
}

message UserResponse {
  string message = 1;
}

message UserListResponse {
  repeated string users = 1;
}

message UpdateMessage{
  string SuccessMessage = 1;
}

message AccountRequest {
  string name = 1;
}

message AccountStatisticsResponse {
  repeated DataPoint data_points = 1;
}

enum StatisticMetric {
  INCOMES_AMOUNT = 0;
  EXPENSES_AMOUNT = 1;
  SAVING_AMOUNT = 2;
}

message StatisticEntry {
  StatisticMetric metric = 1;
  string value = 2;
}

message CurrencyEntry {
  Currency currency = 1;
  string rate = 2; // Represent BigDecimal as string
}

message DataPointId {
  string account = 1;
  string date = 2;
}

message DataPoint {
  DataPointId id = 1;
  repeated Item incomes = 2;
  repeated Item expenses = 3;
  repeated StatisticEntry statistics = 4;
  repeated CurrencyEntry rates = 5;
}

message AccountStatistics {
  repeated Item incomes = 1;
  repeated Item expenses = 2;
  Saving saving = 3;
}

message AccountS {
  repeated Item incomes = 1;
  repeated Item expenses = 2;
  Saving saving = 3;
}

message UpdateAccountRequest {
  string name = 1;
  AccountS account = 2;
}

message Empty{
}

message UpdateAccountResponse {
  string message = 1;
}
